### DISCUSSION TITLE
Which method of obtaining vertex performance is better？

### DISCUSSION THREAD

**Discussion by ouzhou** [gpuweb/gpuweb#4967] [Q&A]
In WebGPU, when drawing multiple rectangles using the draw instance method, there are two ways to read coordinates: 
1) directly passing coordinates through the vertex buffer
2) reading the instance_index and then retrieving from storage buffer.  

Which method performs better?

If the performance of the two methods is not significantly different, considering other factors such as frequent modification of point coordinates , point size, color, etc., it may be more convenient to read from the storage buffer?

in this [demo](https://webgpufundamentals.org/webgpu/lessons/webgpu-points.html)

how about write it  by using storage buffer to receive points position , and use `points[instance_index].xy` to get positon?

![image](https://github.com/user-attachments/assets/bcd78df0-0c9b-4e4e-b3cf-d0f0ca7b5605)

**Comment by kainino0x** [comment]
The term for 2 is "programmable vertex pulling" or "vertex pulling". I'll call 1 "vertex buffers".

Generally, vertex buffers should never be worse than vertex pulling. If, on some GPU, vertex buffers are worse than vertex pulling, then that GPU's driver should just implement vertex buffers using vertex pulling! And in fact some GPUs do implement vertex buffers using vertex pulling (this can be seen in AMD shader disassemblies, IIRC).

OTOH, vertex pulling can definitely be worse for performance on some devices. See here for example:
https://github.com/gpuweb/gpuweb/pull/411#issuecomment-545262832

**Reply by kainino0x** [reply]
Vertex buffers have an inherent latency benefit in that the GPU knows which vertex data should go to which shader invocation from the very beginning, while with vertex pulling it doesn't know until that line of the shader executes - unless the driver statically analyzes the shader and determines that it can prefetch the data.

I should also point out that there are a bunch of vertex formats that can only be used with vertex buffers, and not with vertex pulling (at least not until we have texel buffers: #4912). Those vertex formats are often _smaller_, reducing the necessary memory bandwidth, which can be very significant for performance.

**Comment by greggman** [comment]
Very Old timings

https://github.com/nlguillemot/ProgrammablePulling

https://www.yosoygames.com.ar/wp/2018/03/vertex-formats-part-2-fetch-vs-pull/

Personally I don't think it's worth obsessing one over the other. I'm relatively confident that, for example, not a single [three.js example](https://threejs.org/examples) would show a noticble per different with one technique vs the other. Feel free to prove me wrong. Most projects are not pushing to the limit. If you do want to push the the limit then you'll need to test for every GPU or add some tests to your code or, like most PC games, provide extensive "graphics options" to the user. For me, I'd just use whatever is easier for my particular project. I'd use vertex buffers by default and use storage buffers when I need the flexibility.
